modules/opencascade/opencascade/relax_dynamics_surface.cpp

```diff
@@ -0,0 +1,108 @@
+        {
+            shape_ = real_body_.body_shape_;
+           
+        }
+        //=================================================================================================//
+        void ShapeSurfaceBounding2::update(size_t index_i, Real dt)
+        {
+            pos_[index_i] = shape_->findClosestPoint(pos_[index_i]);
+           
+        }
+        //=================================================================================================//
+        RelaxationStepInnerFirstHalf::
+            RelaxationStepInnerFirstHalf(BaseInnerRelation &inner_relation, bool level_set_correction)
+            : BaseDynamics<void>(inner_relation.getSPHBody()), real_body_(inner_relation.real_body_),
+              inner_relation_(inner_relation)
+        {
+            relaxation_acceleration_inner_ =
+                makeUnique<InteractionDynamics<RelaxationAccelerationInner>>(inner_relation);
+        }
+        //=================================================================================================//
+        void RelaxationStepInnerFirstHalf::exec(Real dt)
+        {
+            real_body_->updateCellLinkedList();
+            inner_relation_.updateConfiguration();
+            relaxation_acceleration_inner_->exec();
+        }
+    
+        //=================================================================================================//
+        RelaxationStepInnerSecondHalf::
+            RelaxationStepInnerSecondHalf(BaseInnerRelation &inner_relation, bool level_set_correction)
+            : BaseDynamics<void>(inner_relation.getSPHBody()), real_body_(inner_relation.real_body_),
+              get_time_step_square_(*real_body_), update_particle_position_(*real_body_),
+              surface_bounding_(*real_body_)
+        {
+        }
+        //=================================================================================================//
+        void RelaxationStepInnerSecondHalf::exec(Real dt)
+        {
+            Real dt_square = get_time_step_square_.exec();
+            update_particle_position_.exec(dt_square);
+            surface_bounding_.exec();
+        }
+ 
+		//=================================================================================================//
+        SurfaceNormalDirection::SurfaceNormalDirection(SPHBody &sph_body)
+            : RelaxDataDelegateSimple(sph_body), LocalDynamics(sph_body), 
+            surface_shape_(DynamicCast<SurfaceShape>(this, sph_body.body_shape_)), 
+            pos_(particles_->pos_), n_(*particles_->getVariableByName<Vecd>("NormalDirection")) {}
+
+        //=================================================================================================//
+        void SurfaceNormalDirection::update(size_t index_i, Real dt)
+        {
+            Extrema_ExtAlgo Algo = Extrema_ExtAlgo_Tree;
+            gp_Vec tangent_u, tangent_v;
+            gp_Vec norm;
+            Vecd normal_direction_;
```

Comment:
normal_direction_ change to normal_direction; "name_ " generally represents class member variable.